src/Simplify.elm

```diff
@@ -6025,6 +6036,156 @@ taskOnErrorChecks checkInfo =
+        , \() -> sequenceOrFirstEmptyChecks taskWithSucceedAsWrap checkInfo
+        ]
+        ()
+
+
+taskSequenceCompositionChecks : CompositionIntoCheckInfo -> Maybe ErrorInfoAndFix
+taskSequenceCompositionChecks checkInfo =
+    mappableSequenceCompositionChecks taskWithSucceedAsWrap checkInfo
+
+
+sequenceOrFirstEmptyChecks :
+    WrapperProperties
+        { otherProperties
+            | empty :
+                { empty
+                    | is : ModuleNameLookupTable -> Node Expression -> Bool
+                    , description : Description
+                }
+        }
+    -> CheckInfo
+    -> Maybe (Error {})
+sequenceOrFirstEmptyChecks emptiable checkInfo =
+    case AstHelpers.getListLiteral checkInfo.firstArg of
+        Just list ->
+            case List.filter (\el -> isNothing (emptiable.wrap.getValue checkInfo.lookupTable el)) list of
+                firstNonWrappedElement :: _ ->
+                    if emptiable.empty.is checkInfo.lookupTable firstNonWrappedElement then
+                        Just
+                            (Rule.errorWithFix
+                                { message = qualifiedToString checkInfo.fn ++ " on a list containing " ++ descriptionForIndefinite emptiable.empty.description ++ " will result in " ++ descriptionForDefinite "the first" emptiable.empty.description
+                                , details = [ "You can replace this call by " ++ descriptionForDefinite "the first" emptiable.empty.description ++ " in the list." ]
+                                }
+                                checkInfo.fnRange
+                                (replaceBySubExpressionFix checkInfo.parentRange firstNonWrappedElement)
+                            )
+
+                    else
+                        Nothing
+
+                [] ->
+                    Nothing
+
+        Nothing ->
+            Nothing
+
+
+wrapperSequenceChecks : WrapperProperties { otherProperties | mapFnName : String } -> CheckInfo -> Maybe (Error {})
+wrapperSequenceChecks wrapper checkInfo =
+    let
+        wrapFn : ( ModuleName, String )
+        wrapFn =
+            ( wrapper.moduleName, wrapper.wrap.fnName )
+    in
+    firstThatConstructsJust
+        [ \() ->
+            callOnEmptyReturnsCheck
+                { on = checkInfo.firstArg
+                , resultAsString =
+                    \res ->
+                        qualifiedToString (qualify wrapFn res)
+                            ++ " []"
+                }
+                listCollection
+                checkInfo
+        , \() ->
+            case AstHelpers.getListSingleton checkInfo.lookupTable checkInfo.firstArg of
+                Just singletonList ->
+                    let
+                        mapFn : ( ModuleName, String )
+                        mapFn =
+                            ( wrapper.moduleName, wrapper.mapFnName )
+
+                        replacement qualifyResources =
```

Comment:
This doesn't have to be a function, right? This could already be a constant that uses `checkInfo` instead of `qualifyResources`.

Reply:
It's used with `defaultQualifyResources` in the error info and with `checkInfo` in the fix.
Adding extra qualification in the error is nice I think? This is what we do for all other errors at least.

Reply:
What I mean is that there is no need to have `replacement` be a function in this context of this function. You could have written.

```elm
replacement =
      qualifiedToString (qualify mapFn checkInfo)
          ++ " "
          ++ qualifiedToString (qualify ( [ "List" ], "singleton" ) checkInfo)
```

I see the value of making this a function if it was possible that we wouldn't need to compute the replacement (based on conditions evaluated later), but in this case there is no more branching that would make this function not be called with `checkInfo`.

Reply:
I think we're both misunderstanding...
The let declared function _is used with different arguments_ in different contexts here.
For the fix, it's always called with `checkInfo`, just like you've pointed out.
But in the error message and error details, you'll find calls to `replacement defaultQualifyResources` which for example return `Task.sequence ...` instead of `sequence ...` even when `sequence` was imported and exposed in the relevant module.

Reply:
Aha! Alright, that makes sense! Sorry and thank you for your patience :sweat_smile: